include/flucoma/clients/nrt/DataSetClient.hpp

```diff
@@ -27,6 +27,28 @@ enum { kName };
 constexpr auto DataSetParams =
     defineParameters(StringParam<Fixed<true>>("name", "Name of the DataSet"));
 
+double distance(FluidTensorView<const double, 1> point1,
+                FluidTensorView<const double, 1> point2)
```

Comment:
This will let you get rid of the temporary buffer for the input point

```suggestion
template<typename T, typename U> 
double distance(FluidTensorView<T, 1> point1,
                FluidTensorView<U, 1> point2)
```

Reply:
is this really a double that is yield from distance, or we do auto like below?

Reply:
also, I sort of get the syntax of template definition first but a little verbose helper would help me understand the consequences to line 244 - I like the elegance but I want to understand the grammar and logic, please

Reply:
Templating on the value types of the `FluidTensorView`s means that we're no longer need to construct temporary buffers of `FluidTensor<double,N>` just to do conversion from `float` – that can happen in `transform_reduce` instead. So on line 244 (and its counterparts) instead of making a new `FluidTensor` we can just take the `FluidTensorView<float,1>` returned by `samples()` and use that directly.

Reply:
ok I've implemented it all and pushed. but I still have a question about the templating. SHould we not force the templates to have the same typename T for both inputs to the function? or we just go super permissing like this and expect the first compile pass to check that T and U can actually sum to the difference of 2 vectors and generate double out of that?

Reply:
The point is that we need to be able to handle them being different (specifically `float` in one, `double` in other). We could add some (super-ugly before C++20) restrictions to insist that both `T` and `U` are at least numeric, or even floating point, which would be worth it if this function were made to be called more widely. 

For our actual use cases, there's no problem: `transform_reduce` will be using `double` internally (inferred from the type of the initialization value we pass), and if the compiler can't work out how to do `x - y` for each of the points, it will moan.